perceval/backends/gerrit.py

```diff
@@ -0,0 +1,308 @@
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
+#
+# Authors:
+#   Alvaro del Castillo San Felix <acs@bitergia.com>
+#
+
+'''Gerrit backend for Perseval'''
+
+
+from datetime import datetime
+import json
+import logging
+import os.path
+import re
+import subprocess
+from time import time
+
+from ..backend import Backend, BackendCommand
+from ..cache import Cache
+from ..errors import BackendError, CacheError
+from ..utils import DEFAULT_DATETIME, str_to_datetime
+
+
+class Gerrit(Backend):
+
+    name = "gerrit"
+
+    def __init__(self, user=None, url=None, nreviews=None,
+                 cache=None, **nouse):
+        super().__init__(cache=cache)
+        self.repository = url
+        self.nreviews = nreviews
+        self.last_item = None  # Start item for next iteration
+        self.more_updates = True  # To check if reviews are updates
+        self.number_results = self.nreviews  # Check for more items
+        self.client = GerritClient(self.repository, user, nreviews)
+
+    def get_id(self):
+        ''' Return gerrit unique identifier '''
+
+        return self.repository
+
+    def get_field_unique_id(self):
+        return "id"
+
+    def fetch(self, from_date=DEFAULT_DATETIME):
+
+        self._purge_cache_queue()
+
+        reviews = self.get_reviews(from_date)
+        self._push_cache_queue(reviews)
+
+        while reviews:
+            issue = reviews.pop(0)
+            yield issue
+
+            if not reviews:
+                reviews = self.get_reviews(from_date)
+                self._push_cache_queue(reviews)
+                self._flush_cache_queue()
+
+    def fetch_from_cache(self):
+        """Fetch the bugs from the cache.
+
+        It returns the bugs stored in the cache object provided during
+        the initialization of the object. If this method is called but
+        no cache object was provided, the method will raise a `CacheError`
+        exception.
+
+        :returns: a generator of items
+
+        :raises CacheError: raised when an error occurs accesing the
+            cache
+        """
+        if not self.cache:
+            raise CacheError(cause="cache instance was not provided")
+
+        cache_items = self.cache.retrieve()
+
+        while True:
+            try:
+                raw_items = next(cache_items)
+                if raw_items is None:
+                    break
+            except StopIteration:
+                break
+
+            for item in raw_items:
+                yield item
+
+    def get_reviews(self, from_date=None):
+        """ Get all reviews from repository """
+
+        reviews = []
+
+        if self.number_results < self.nreviews or not self.more_updates:
+            # No more reviews after last iteration
+            return reviews
+
+        task_init = time()
+        raw_data = self.client.get_items(self.last_item)
```

Comment:
@sduenas you said that retrieval and parsing logic are mixed here. Not sure. The retrieval logic is inside self.client.get_items. We get the raw data here. And then the parsing logic starts. But here, the iteration for the retrieval of reviews is also here. And probably this is the error. get_reviews should read from a generator inside GerritClient and then parse tha raw data that comes from this generator. 

The same approach that should be done in the cache once the code is fixed to store real raw data in the cache.